game_catalog/views.py

```diff
@@ -1,3 +1,109 @@
-from django.shortcuts import render
+from rest_framework import viewsets, status, generics
+from rest_framework.decorators import action
+from rest_framework.generics import CreateAPIView
+from rest_framework.permissions import IsAuthenticated, IsAdminUser, AllowAny
+from rest_framework.response import Response
 
-# Create your views here.
+from .custom_permissions import (
+    IsAdminOrReadOnly,
+    IsOwnerOrAdmin,
+    IsAuthenticatedOrReadOnly,
+)
+
+from .models import Genre, Studio, Game, Comment, CustomUser
+from .serializers import (
+    GenreSerializer,
+    StudioSerializer,
+    GameSerializer,
+    GameWriteSerializer,
+    CommentSerializer,
+    UserExtendedInfoSerializer,
+    UserShortInfoSerializer,
+    RegisterSerializer,
+)
+
+
+class RegisterView(CreateAPIView):
+    permission_classes = [AllowAny]
+    serializer_class = RegisterSerializer
+
+    def perform_create(self, serializer):
+        user = serializer.save()
+        user.set_password(user.password)
+        user.save()
+        return user
+
+
+class GenreViewSet(viewsets.ModelViewSet):
+    queryset = Genre.objects.all()
+    serializer_class = GenreSerializer
+    permission_classes = [IsAdminOrReadOnly]
+
+
+class StudioViewSet(viewsets.ModelViewSet):
+    queryset = Studio.objects.all()
+    serializer_class = StudioSerializer
+    permission_classes = [IsAdminOrReadOnly]
+
+
+class GameViewSet(viewsets.ModelViewSet):
+    queryset = Game.objects.select_related("studio").prefetch_related("genre")
+    permission_classes = [IsAdminOrReadOnly]
+
+    def get_serializer_class(self):
+        if self.action in ["create", "update", "partial_update"]:
+            return GameWriteSerializer
+        return GameSerializer
+
+    @action(detail=True, methods=["post"], permission_classes=[IsAuthenticated])
+    def toggle_favorite(self, request, pk=None):
+        game = self.get_object()
+        user = request.user
+
+        if game in user.favorite_games.all():
+            user.favorite_games.remove(game)
+            message = "Game removed from favorites."
+        else:
+            user.favorite_games.add(game)
+            message = "Game added to favorites."
+
+        serializer = UserShortInfoSerializer(user)
+        return Response({"message": message, "user": serializer.data}, status=status.HTTP_200_OK)
```

Comment:
Add an `is_favorite` Method or Property on the Model:
 - Centralize the logic for checking whether a game is a favorite in the User model.
    Example:

```python
    class User(models.Model):
        def is_favorite(self, game):
            return self.favorite_games.filter(pk=game.pk).exists()
```

Extract the Toggle Logic to a Global Action:

 - Create a reusable function, actions.toggle_favorite, to handle toggling logic. This function should:
        - Accept a user and a model instance (e.g., a game).
        - Use the is_favorite method to check the current state.
        - Add/remove the instance from the user's favorites accordingly.
        - Return the updated state or any relevant data for further processing.
        
    Example:
```python
    def toggle_favorite(user, instance):
        if user.is_favorite(instance):
            user.favorite_games.remove(instance)
        else:
            user.favorite_games.add(instance)
```
 - Refactor the GameViewSet.toggle_favorite Method:

    - Replace inline logic with the call to the global actions.toggle_favorite function.
    - Remove the message field from the response since this is an API call, and the updated state is sufficient.
    Example:
```python
        @action(detail=True, methods=["post"], permission_classes=[IsAuthenticated])
        def toggle_favorite(self, request, pk=None):
            game = self.get_object()
            user = request.user

            actions.toggle_favorite(user, game)

            return Response({
                "is_favorite": user.is_favorite(game),
                "user": UserShortInfoSerializer(user).data,
            }, status=status.HTTP_200_OK)
```
- Follow Dependency Injection Principles:  
         - Ensure actions.toggle_favorite is decoupled from specific model implementations. It should handle any model with a similar relationship structure (e.g., favorites).
         - Avoid hardcoding logic for Game or favorite_games directly, allowing broader reuse.

 - Expected Outcomes:

    - Global Reusability: The toggle_favorite logic becomes a generic, reusable utility for any favorite-like relationship.
    - SRP Alignment: The model (is_favorite) handles the data-specific logic, the global action focuses on toggling, and the view manages HTTP concerns.
    - Improved Clarity: Each layer in the application has a distinct responsibility, making the codebase easier to maintain, test, and extend.

Reply:
- Added `is_favorite `method on `CustomUser `model
- Created reusable `actions.toggle_favorite` function
- Refactored `GameViewSet.toggle_favorite` to use `actions.toggle_favorite`

b2372dee420981021b6d70bc5f691d00d6de2e93